x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityMigrationExecutor.java

```diff
@@ -0,0 +1,117 @@
+import java.util.concurrent.Executor;
+
+import static org.elasticsearch.xpack.security.support.SecurityIndexManager.loadIndexMappingVersion;
+
+public class SecurityMigrationExecutor extends PersistentTasksExecutor<SecurityMigrationTaskParams> {
+
+    private static final Logger logger = LogManager.getLogger(SecurityMigrationExecutor.class);
+    private final SecuritySystemIndices securitySystemIndices;
+    private final Client client;
+    private final FeatureService featureService;
+    private final ClusterService clusterService;
+    private final TreeMap<Integer, SecurityMigrations.SecurityMigration> securityMigrationByMigrationVersion;
+
+    public SecurityMigrationExecutor(
+        String taskName,
+        Executor executor,
+        SecuritySystemIndices securitySystemIndices,
+        Client client,
+        FeatureService featureService,
+        ClusterService clusterService,
+        TreeMap<Integer, SecurityMigrations.SecurityMigration> securityMigrationByMigrationVersion
+    ) {
+        super(taskName, executor);
+        this.securitySystemIndices = securitySystemIndices;
+        this.client = client;
+        this.featureService = featureService;
+        this.clusterService = clusterService;
+        this.securityMigrationByMigrationVersion = securityMigrationByMigrationVersion;
+    }
+
+    @Override
+    protected void nodeOperation(AllocatedPersistentTask task, SecurityMigrationTaskParams params, PersistentTaskState state) {
+        applyOutstandingMigrations(params.getMigrationVersion(), ActionListener.wrap((res) -> task.markAsCompleted(), (exception) -> {
+            logger.warn("Security migration failed: " + exception);
+            task.markAsFailed(exception);
+        }));
+    }
+
+    private void applyOutstandingMigrations(int currentMigrationVersion, ActionListener<Void> listener) {
+        if (securityMigrationByMigrationVersion.lastKey() > currentMigrationVersion) {
+            Map.Entry<Integer, SecurityMigrations.SecurityMigration> migrationEntry = securityMigrationByMigrationVersion.higherEntry(
+                currentMigrationVersion
+            );
+
+            int currentMappingsVersion = loadIndexMappingVersion(
+                securitySystemIndices.getMainIndexManager().aliasName(),
+                clusterService.state()
+            );
+
+            SecurityMigrations.SecurityMigration migration = migrationEntry.getValue();
+            // If all nodes can't support feature, or the migration needs a more recent mapping version applied, stop migration
+            if (migration.nodeFeaturesRequired()
+                .stream()
+                .anyMatch((feature) -> featureService.clusterHasFeature(clusterService.state(), feature) == false)
+                || migration.minMappingVersion() > currentMappingsVersion) {
+                logger.info("Security migrations applied until version: [" + migrationEntry.getKey() + "]");
+                listener.onResponse(null);
```

Comment:
nit: you need a return after this listener#onResponse.

Reply:
Thanks! This actually caused all migration to run, even though they were not eligible, so my unit test failed.